tests/templates/test_layers.py

```diff
@@ -642,3 +643,146 @@ def circuit(weights):
+        and order, the wires each operation acts on and the correct use of parameters
+        in the circuit."""
+        layers = 2
+        qubits = 4
+
+        weights = np.random.normal(0, 2 * np.pi, (layers, 2 * qubits - 1))
+
+        n_gates = 1 + (qubits + (qubits - 1) * 3) * layers
+
+        exp_gates = (
+            [qml.RZ] * qubits + ([qml.CNOT] + [qml.CRX] + [qml.CNOT]) * (qubits - 1)
+        ) * layers
+
+        with pennylane._queuing.OperationRecorder() as rec:
+            ParticleConservingU2(weights, wires=range(qubits), init_state=init_state)
+
+            # number of gates
+            assert len(rec.queue) == n_gates
+
+            # initialization
+            assert isinstance(rec.queue[0], qml.BasisState)
+
+            # order of gates
+            for op1, op2 in zip(rec.queue[1:], exp_gates):
+                assert isinstance(op1, op2)
+
+            # gate parameter
+            params = np.array(
+                [
+                    rec.queue[i].parameters
+                    for i in range(1, n_gates)
+                    if rec.queue[i].parameters != []
+                ]
+            )
+            weights[:, qubits:] = weights[:, qubits:] * 2
+            assert np.allclose(params.flatten(), weights.flatten())
+
+            # gate wires
+            wires = Wires(range(qubits))
+            nm_wires = [wires.subset([l, l + 1]) for l in range(0, qubits - 1, 2)]
+            nm_wires += [wires.subset([l, l + 1]) for l in range(1, qubits - 1, 2)]
+
+            exp_wires = []
+            for _ in range(layers):
+                for i in range(qubits):
+                    exp_wires.append(wires.subset([i]))
+                for j in nm_wires:
+                    exp_wires.append(j)
+                    exp_wires.append(j[::-1])
+                    exp_wires.append(j)
+
+            res_wires = [rec.queue[i]._wires for i in range(1, n_gates)]
+
+            assert res_wires == exp_wires
+
+    @pytest.mark.parametrize(
+        ("weights", "wires", "init_state", "msg_match"),
+        [
+            (
+                np.array([[-0.080, 2.629, -0.710, 5.383, 0.646, -2.872, -3.856]]),
+                [0],
+                np.array([1, 1, 0, 0]),
+                "This template requires the number of qubits to be >= 2",
+            ),
+            (
+                np.array([[-0.080, 2.629, -0.710, 5.383]]),
+                [0, 1, 2, 3],
+                np.array([1, 1, 0, 0]),
+                "'weights' must be of shape",
+            ),
+            (
+                np.array(
+                    [
+                        [-0.080, 2.629, -0.710, 5.383, 0.646, -2.872],
+                        [-0.080, 2.629, -0.710, 5.383, 0.646, -2.872],
+                    ]
+                ),
+                [0, 1, 2, 3],
+                np.array([1, 1, 0, 0]),
+                "'weights' must be of shape",
+            ),
+        ],
+    )
```

Comment:
this is minor: ``init_state`` is fixed so it can be moved from the list of parameters to the body of the function.